### PR TITLE
[fancode] support fancodeLive and fixes

### PR DESCRIPTION
**Pull request by zenerdi0de**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adds support for fancode live streams, and fixes vod extraction (brightcove user id is no longer available in the file name).

Since it uses openid to login, use `refresh` as the username and the `__refreshToken` as the password to login instead.

### CODE REVIEW

File: yt_dlp/extractor/fancode.py [hunk 1]
```diff
@@ -38,16 +39,54 @@ class FancodeVodIE(InfoExtractor):
...
+
+            token_json = self.download_gql('refresh token', data, "Getting the Access token")
+            self._ACCESS_TOKEN = try_get(token_json, lambda x: x['data']['refreshToken']['accessToken'])
+            if self._ACCESS_TOKEN is None:
+                raise ExtractorError('Failed to get Access token')
+            else:
+                self._REFRESH_TOKEN = password
+                self.headers.update({'Authorization': 'Bearer %s' % self._ACCESS_TOKEN})
+        else:
+            pass
```
Comment (pukkandan) [inline yt_dlp/extractor/fancode.py:73]: If some other username was given, warn the user about how to actually use the option

File: yt_dlp/extractor/fancode.py [hunk 2]
```diff
@@ -38,16 +39,54 @@ class FancodeVodIE(InfoExtractor):
...
+                "variables":{
+                    "refreshToken":"%s"
+                },
+                "operationName":"RefreshToken"
+            }''' % password
+
+            token_json = self.download_gql('refresh token', data, "Getting the Access token")
+            self._ACCESS_TOKEN = try_get(token_json, lambda x: x['data']['refreshToken']['accessToken'])
+            if self._ACCESS_TOKEN is None:
+                raise ExtractorError('Failed to get Access token')
```
Comment (pukkandan) [inline yt_dlp/extractor/fancode.py:68]: should just be a warning

File: yt_dlp/extractor/fancode.py [hunk 3]
```diff
@@ -38,16 +39,54 @@ class FancodeVodIE(InfoExtractor):
...
+            else:
+                self._REFRESH_TOKEN = password
```
Comment (pukkandan) [inline yt_dlp/extractor/fancode.py:70]: why save this in the class?

File: yt_dlp/extractor/fancode.py [hunk 4]
```diff
@@ -38,16 +39,54 @@ class FancodeVodIE(InfoExtractor):
...
+    _NETRC_MACHINE = 'fancode'
+
+    headers = {
+        'content-type': 'application/json',
+        'origin': 'https://fancode.com',
+        'referer': 'https://fancode.com',
+    }
+
+    def _login(self):
+        # Access tokens are shortlived, so get them using the refresh token.
```
Comment (pukkandan) [inline yt_dlp/extractor/fancode.py:53]: If user provides cookies, cant refresh token be extracted from there?

File: yt_dlp/extractor/fancode.py [hunk 5]
```diff
@@ -74,8 +107,12 @@ def _real_extract(self, url):
             raise ExtractorError('Unable to extract brightcove Video ID')
 
         is_premium = media.get('isPremium')
-        if is_premium:
-            self.report_warning('this video requires a premium account', video_id)
+        is_available = media.get('isUserEntitled')
+
+        if is_premium and self._ACCESS_TOKEN is None:
+            self.raise_login_required()
```
Comment (pukkandan) [inline yt_dlp/extractor/fancode.py:113]: Do cookies work for login? If not, use:
```suggestion
            self.raise_login_required(method='password')
```
Same everywhere else

File: yt_dlp/extractor/fancode.py (hunk 4 above)
Comment (zenerdi0de) [reply yt_dlp/extractor/fancode.py:53]: I didn't find any way to do that. They only seem to set when a user first logs in.

File: yt_dlp/extractor/fancode.py (hunk 5 above)
Comment (pukkandan) [reply yt_dlp/extractor/fancode.py:113]: Actually, don't do this. Its better to modify the `raise_login_required` function to allow a custom message as `method` and tell the user to use the `-u "refresh" -p <refresh_token>`

**Pull request merged**